### PR TITLE
Fixed peer that were stuck

### PR DESCRIPTION
**Pull request by MithrilMan**
moved the peersbyid.add call earlier
adjusted the peer console dump (added cacheHeaders information and compressed the Inbound/Outbound direction

UPDATE:
fixed a nasty bug that was causing peers to stuck (and potentially stuck node too)

### CODE REVIEW

File: src/Stratis.Bitcoin/Consensus/ConsensusManager.cs [hunk 1]
```diff
@@ -257,6 +257,12 @@ public ConnectNewHeadersResult HeadersPresented(INetworkPeer peer, List<BlockHea
 
                 connectNewHeadersResult = this.chainedHeaderTree.ConnectNewHeaders(peerId, headers);
 
+                if (!this.peersByPeerId.ContainsKey(peerId))
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: The position in the locked code segment that this is placed seems a bit arbitrary (both doing it last and moving it higher up). What was the reasoning behind putting it after `ConnectNewHeaders` in particular instead of, say, just before it?

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: @dangershony didn't wanted to put it before because in case `ConnectNewHeaders` throws the peer will be disconnected, so no need to att to the peersByPeerId dictionary.

Probably it's right because we may incur in a deadlock if we put it before ConnectNewHeaders because it causes peer to disconnect.

Anyway there are more things to investigate about peersByPeerId because it seems this gets populated only for peers supporting PH (after last checkpoint) and this may cause problems

**Comment by MithrilMan** [comment]
Don't merge yet, I'll fix 3168 here too

**Comment by MithrilMan** [comment]
fix applied, I need to sync from scratch to confirm problem is fixed (tomorrow)

**Comment by MithrilMan** [comment]
Ok fixed another important bug about "insufficientInfo" exception, now peers aren't stuck anymore and all seems consistent even at console level (finally all these peers at same r/s height)

Of course legacy peers (StratisX and pre PH SBFN) report latest received header as checkpoint height because they doesn't collaborate after that point (that's by design)

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 2]
```diff
@@ -112,7 +112,18 @@ public async Task<ConnectNewHeadersResult> ConsensusTipChangedAsync()
                     this.BestReceivedTip = result.Consumed;
                     this.UpdateBestSentHeader(this.BestReceivedTip);
 
-                    int consumedCount = this.cachedHeaders.IndexOf(result.Consumed.Header) + 1;
+                    uint256 consumedHeaderHash = result.Consumed.Header.GetHash();
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: Is this change necessary?

We probably also need to make this in the ProcessesHeadrs method

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 3]
```diff
@@ -659,5 +670,10 @@ public override object Clone()
         {
             return new ConsensusManagerBehavior(this.chain, this.initialBlockDownloadState, this.consensusManager, this.peerBanning, this.loggerFactory);
         }
+
+        internal int GetCachedItemsCount()
+        {
+            return this.cachedHeaders.Count;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: hmm what about a lock? 

Maybe for taking the count it snot a big deal.

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs (hunk 2 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: of course it is because it was comparing two different instances of BlockHeader and it returned -1 even if the header was in the list, so I had to compare their hash

Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: I could have used FindIndex Linq but I wanted a backward loop so I did an explicit loop

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 4]
```diff
@@ -654,8 +654,13 @@ public ConnectNewHeadersResult ConnectNewHeaders(int networkPeerId, List<BlockHe
 
             if (insufficientInfo)
             {
+                if (this.TryFindLastValidatedHeader(headers, out ChainedHeader lastAlreadyValidatedHeader))
+                {
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Maybe add a trace log in the if condition?

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 5]
```diff
@@ -112,7 +112,18 @@ public async Task<ConnectNewHeadersResult> ConsensusTipChangedAsync()
                     this.BestReceivedTip = result.Consumed;
                     this.UpdateBestSentHeader(this.BestReceivedTip);
 
-                    int consumedCount = this.cachedHeaders.IndexOf(result.Consumed.Header) + 1;
+                    uint256 consumedHeaderHash = result.Consumed.Header.GetHash();
+                    int consumedCount = 0;
+
+                    for (int i = this.cachedHeaders.Count - 1; i >= 0; i--)
+                    {
+                        if (this.cachedHeaders[i].GetHash() == consumedHeaderHash)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: Just double check that the headers are cached then this is ok

Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: and also as discussed move it to its own method and apply in the `ProcessHeader` method

File: src/Stratis.Bitcoin/Consensus/ConsensusManager.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: @dangershony @MithrilMan I'm also sketchy on why this has been moved up? Care to elaborate?

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 6]
```diff
@@ -234,8 +234,9 @@ private void AddComponentStats(StringBuilder builder)
             {
                 var chainHeadersBehavior = peer.Behavior<ConsensusManagerBehavior>();
 
-                string peerHeights = $"(r/s):{(chainHeadersBehavior.BestReceivedTip != null ? chainHeadersBehavior.BestReceivedTip.Height.ToString() : peer.PeerVersion?.StartHeight + "*" ?? "-")}";
+                string peerHeights = $"(r/s/c):{(chainHeadersBehavior.BestReceivedTip != null ? chainHeadersBehavior.BestReceivedTip.Height.ToString() : peer.PeerVersion?.StartHeight + "*" ?? "-")}";
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: `?? "-"` is unreachable
same on the line below

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 7]
```diff
@@ -234,8 +234,9 @@ private void AddComponentStats(StringBuilder builder)
...
                 peerHeights += $"/{(chainHeadersBehavior.BestSentHeader != null ? chainHeadersBehavior.BestSentHeader.Height.ToString() : peer.PeerVersion?.StartHeight + "*" ?? "-")}";
+                peerHeights += $"/{chainHeadersBehavior.GetCachedItemsCount()}";
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: instead of 
```
a = "q"
a += "w"
```

better to do 

```
a = "q" +
      "w"
```

mostly for consistency

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 8]
```diff
@@ -1103,6 +1109,30 @@ private bool TryFindNewHeaderIndex(List<BlockHeader> headers, out int newHeaderI
...
         }
 
+        /// <summary>
+        /// Find the last validated <see cref="ChainedHeader"/> header in the given list of <see cref="headers"/>.
+        /// </summary>
+        /// <remarks>Headers are supposed to be consecutive and sorted by height</remarks>
+        private bool TryFindLastValidatedHeader(List<BlockHeader> headers, out ChainedHeader lastValidatedHeader)
+        {
+            lastValidatedHeader = null;
+            for (int i = 0; i < headers.Count; i++)
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: can be replaced with foreach

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 9]
```diff
@@ -1103,6 +1109,30 @@ private bool TryFindNewHeaderIndex(List<BlockHeader> headers, out int newHeaderI
...
+            {
+                uint256 currentBlockHash = headers[i].GetHash();
+                if (this.chainedHeadersByHash.TryGetValue(currentBlockHash, out ChainedHeader currentChainedHeader))
+                {
+                    lastValidatedHeader = currentChainedHeader;
+                }
+                else
+                {
+                    // stop at the first header not found.
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: `Stop`

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 10]
```diff
@@ -654,8 +654,14 @@ public ConnectNewHeadersResult ConnectNewHeaders(int networkPeerId, List<BlockHe
 
             if (insufficientInfo)
             {
+                if (this.TryFindLastValidatedHeader(headers, out ChainedHeader lastAlreadyValidatedHeader))
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: This looks correct, gj

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 11]
```diff
@@ -659,5 +660,32 @@ public override object Clone()
...
+        }
+
+        /// <summary>
+        /// Gets the count of consumed headers in a <paramref name="headers"/> list, giving a <paramref name="consumedHeader"/> reference.
+        /// All items up to <paramref name="consumedHeader"/> are considered consumed.
+        /// </summary>
+        /// <param name="headers">List of headers to use to get the consumed count.</param>
+        /// <param name="consumedHeader">The consumed header reference.</param>
+        /// <returns>The number of consumed cached items.</returns>
+        private int GetConsumedHeadersCountLockedAsync(List<BlockHeader> headers, BlockHeader consumedHeader)
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: its not async

Comment (noescape00) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: when there is locked in the name there should be a comment referencing by which lock it should be locked

And I don't think it should be locked

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 6 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I preserved the previous code. If you are sure that PeerVersion can never be null than we can remove it

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 7 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: consistency with what?
I saw the code was already doing a += (L238) and so I used that syntax for consistency too
It's ok for me to fix it if it break the way we are used to handle string composition (even if tbh I'd use a StringBuilder to do those concatenations)

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 6 above)
Comment (noescape00) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: it maybe can. but even if it is `?? "-"` is still unreachable

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs (hunk 11 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: both method calls are under lock, so I added the Lock to the name even because one of the call need cachedItems that must be under lock (initially the method didn't had the headers argument but was using only cachedHeaders but Dan asked to be used on both places, so I had to add an headers argument to be able to use it in the part of code that doesn't use cachedHeaders.

I'm not sure what to do in this case, maybe It's ok to remove the Locked at this point

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 6 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I agree it will never be reached because you always have the `+ "*"`  I will fix it

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 7 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: small change to make, Fixed.

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs (hunk 10 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Thats the main fix indeed.

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs (hunk 8 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Bot big deal I leave it as is

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs (hunk 9 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Fixed

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 7 above)
Comment (noescape00) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: >consistency with what?

with the majority of the codebase. 
but it's up to you to fix it or not, it's not a policy or smth

File: src/Stratis.Bitcoin/Consensus/ConsensusManager.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManager.cs]: Its a bit long but here goes:

The bellow code is valid scenarios of the flow 
 
From CM.HeadersPresented method
```
                if (connectNewHeadersResult == null)
                {
                    this.logger.LogTrace("(-)[NO_HEADERS_CONNECTED]:null");
                    return null;
                }

                if (connectNewHeadersResult.Consumed == null)
                {
                    this.logger.LogTrace("(-)[NOTHING_CONSUMED]");
                    return connectNewHeadersResult;
                }
```
If both the above conditions are true this means insufficient info to validate a proven header.
This will often happen when the first header is inefficient then  `connectNewHeadersResult.Consumed` will be null (all though @MithrilMan also fixed another bug where we would always report it as null, see the changes to CHT)

If `Consumed == null` causes CMB to do this (basically headers are added to the cache)

```
                if (result.Consumed == null)
                {
                    this.cachedHeaders.AddRange(headers);
                    this.logger.LogDebug("All {0} items were not consumed and added to cache.", headers.Count);

                    return;
                }
```

But this issue is actually reflected later in the method `NotifyBehaviorsOnConsensusTipChangedAsync` this code:
```
     lock (this.peerLock)
            {
                foreach (INetworkPeer peer in this.peersByPeerId.Values)
                    behaviors.Add(peer.Behavior<ConsensusManagerBehavior>());
            }
```

That method will trigger when consensus tip changed and should notify the peer to send the next batch of headers (or actually its proven headers) that are waiting in cache to be processed (because we could not connect them earlier due to missing UTXOs). but the peer was never added to this list `this.peersByPeerId`

So the bug that this change fixes (because the fix to CHT will fix the scenario where any of the headers had insufficient info) is that if we connected to a peer and the first headers it presents could not be connected the peer would be stuck because we would not notify the peer's behaviour to try presenting those headers again when consensus advances.

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs (hunk 11 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: In one case of this method call we use `this.cachedHeaders` which should be called under a lock and in both cases we are anyway the `asyncLock` so I'll remove the `async` and keep the `lock`.

Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: Actually I will remove the `Locked`  this is up to the caller and does not make sense in this context

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 6 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: ah ok I see what you mean
